app/cdap/components/GridTable/components/NestedMenu/index.tsx

```diff
@@ -0,0 +1,149 @@
+import _ from 'lodash';
+import React, { useState } from 'react';
+
+export default function({
+  menuOptions,
+  submitMenuOption,
+  columnType,
+  title,
+  anchorElement,
+  setAnchorElement,
+  menuToggleHandler,
+}: INestedMenuProps) {
+  const [menuComponentOptions, setMenuComponentOptions] = useState<IMenuItem[][]>([]);
+  const classes = useNestedMenuStyles();
+
+  const handleMenuClick = (
+    event: React.MouseEvent<HTMLLIElement, MouseEvent>,
+    menuItem: IMenuItem,
+    origin?: string
+  ) => {
+    event.preventDefault();
+    event.stopPropagation();
+    if (origin === 'parentMenu') {
+      if (menuItem.hasOwnProperty('options') && menuItem?.options?.length > 0) {
+        const updatedAnchors = anchorElement.splice(1, 0, event.currentTarget);
+        setAnchorElement((prev) => anchorElement);
+        setMenuComponentOptions([menuItem?.options]);
+      } else {
+        submitMenuOption(menuItem.value, menuItem.supportedDataType);
+        setAnchorElement(null);
+        menuToggleHandler(title);
+      }
+    } else {
+      if (menuItem.hasOwnProperty('options') && menuItem?.options?.length > 0) {
+        let referenceIndex = -1;
+        menuComponentOptions.forEach((eachMenuOptions, menuOptionsIndex) => {
+          eachMenuOptions.forEach((eachOption, optionsIndex) => {
+            if (eachOption.value === menuItem.value) {
+              referenceIndex = menuOptionsIndex;
+            }
+          });
+        });
+        if (referenceIndex >= 0 && !anchorElement.includes(event.currentTarget)) {
+          const insertPosition = referenceIndex + 2;
+          anchorElement.splice(insertPosition, 0, event.currentTarget);
+          const updatedAnchors = anchorElement.slice(0, insertPosition + 1);
+          menuComponentOptions.splice(updatedAnchors.length - 2, 0, menuItem?.options);
+          setMenuComponentOptions((prev) =>
+            prev.length ? menuComponentOptions : [menuItem?.options]
+          );
+          setAnchorElement(updatedAnchors);
+        } else if (anchorElement.includes(event.currentTarget)) {
+          const currentTargetIndex = _.findIndex(
+            anchorElement,
+            (anchor) => anchor === event.currentTarget
+          );
+          menuComponentOptions.splice(currentTargetIndex, 0, menuItem?.options);
+          setAnchorElement((prev) => prev.slice(0, currentTargetIndex + 1));
+          setMenuComponentOptions((prev) => menuComponentOptions.slice(0, currentTargetIndex + 1));
+        } else {
+          setAnchorElement((prev) => [...prev, event.currentTarget]);
+          setMenuComponentOptions((prev) =>
+            prev.length ? [...prev, menuItem?.options] : [menuItem?.options]
+          );
+        }
+      } else {
+        submitMenuOption(menuItem.value, menuItem.supportedDataType);
+        setAnchorElement(null);
+        menuToggleHandler(title);
+      }
+    }
+  };
+  return (
+    <>
+      <Menu
```

Comment:
These nested menu options and the menu seem like they've been overgeneralized. This code is hard to follow and it seems like instead of writing these giant menu components and are generalized it could be made a lot simpler reusing a couple smaller components. 

Do the menu options change dynamically? If not I don't think it should be made like this.

Reply:
Actually now that I look at it again - the menu only goes one level deep for each of these? Maybe just with more comments this can be fine.

Reply:
Added comments in the file